src/Build/Logging/BinaryLogger/BinaryLogReplayEventSource.cs

```diff
@@ -10,65 +10,54 @@ namespace Microsoft.Build.Logging
                 {
-                    var text = ResourceUtilities.FormatResourceString("UnsupportedLogFileFormat", fileFormatVersion, BinaryLogger.FileFormatVersion);
-                    throw new NotSupportedException(text);
-                }
+                    var gzipStream = new GZipStream(stream, CompressionMode.Decompress, leaveOpen: true);
+                    var binaryReader = new BinaryReader(gzipStream);
 
-                var reader = new BuildEventArgsReader(binaryReader);
-                while (true)
-                {
-                    BuildEventArgs instance = null;
+                    int fileFormatVersion = binaryReader.ReadInt32();
 
-                    try
-                    {
-                        instance = reader.Read();
-                    }
-                    catch (Exception ex)
+                    // the log file is written using a newer version of file format
+                    // that we don't know how to read
+                    if (fileFormatVersion > BinaryLogger.FileFormatVersion)
                     {
-                        string code;
-                        string helpKeyword;
-                        var text = ResourceUtilities.FormatResourceString(out code, out helpKeyword, "InvalidLogFileFormat", ex.Message);
-                        var message = new BuildErrorEventArgs(
-                            subcategory: "",
-                            code: code,
-                            file: sourceFilePath,
-                            lineNumber: 0,
-                            columnNumber: 0,
-                            endLineNumber: 0,
-                            endColumnNumber: 0,
-                            message: text,
-                            helpKeyword: helpKeyword,
-                            senderName: "MSBuild");
-                        Dispatch(message);
+                        var text = ResourceUtilities.FormatResourceString("UnsupportedLogFileFormat", fileFormatVersion, BinaryLogger.FileFormatVersion);
+                        throw new NotSupportedException(text);
                     }
 
-                    if (instance == null)
+                    var reader = new BuildEventArgsReader(binaryReader);
+                    while (true)
                     {
-                        break;
-                    }
+                        BuildEventArgs instance = null;
+
+                        instance = reader.Read();
+                        if (instance == null)
+                        {
+                            break;
+                        }
 
-                    Dispatch(instance);
+                        Dispatch(instance);
+                    }
                 }
             }
+            catch (Exception ex)
+            {
+                string code;
+                string helpKeyword;
+                var text = ResourceUtilities.FormatResourceString(out code, out helpKeyword, "InvalidLogFileFormat", ex.Message);
```

Comment:
there appears to be a FormatResourceString overload that does not have the out params